functions.sql

```diff
@@ -0,0 +1,92 @@
+/* Additional database functions for openstreetmap-carto */
+
+/* Access functions below adapted from https://github.com/imagico/osm-carto-alternative-colors/tree/591c861112b4e5d44badd108f4cd1409146bca0b/sql/roads.sql */
+
+
+/* Simplified 'yes', 'restricted', 'no', NULL scale for access restriction 
+  'no' is returned if the rendering for highway category does not support 'destination'.
+  NULL is functionally equivalent to 'yes', but indicates the absence of a restriction 
+  rather than a positive access = yes */
+CREATE OR REPLACE FUNCTION carto_int_access(primary_mode text, accesstag text)
+	RETURNS text
+	LANGUAGE SQL
+	IMMUTABLE PARALLEL SAFE
+AS $$
+SELECT
+	CASE
+	WHEN accesstag IN ('yes', 'designated', 'permissive', 'customers') THEN 'yes'
+	WHEN accesstag IN ('destination',  'delivery', 'permit') THEN
+		CASE WHEN primary_mode IN ('motorcar', 'pedestrian') THEN 'restricted' ELSE 'no' END
```

Comment:
> You here distinguish between a two class and a three class access restriction model based on the primary mode - while we make this distinction based on road class. While this might not be significant in the exact design model you propose, it would be in many styling variation (like for example if you render `highway=track` with primary mode `motorcar`).

Yes and no. I renamed the functions to `carto_highway_primary_mode` etc. and returned a "primary mode" for clarity, whereas it's really an "access type", so you could classify into `motorcar_track` if you wanted to have a 2 class access style for track. I could change it back to say `access_mode`. I was keen not to "reparse" all the way from highway to avoid repetition and improve efficiency.

Reply:
Primary mode of use and road class are well defined concepts - you will need to explain what you exactly mean by "access type".

Reply:
Fundamentally the "access category" is a data abstraction that keeps the code simple and flexible. A given way is classified into an access category, which mostly depends on the highway type, but for `highway=path` also depends on access tags. All the decisions can then be made on the "access category", e.g. this determines which mode-specific access tags are to be used, but it is also used to distinguish between 2-state and 3-states access renders. So it is a single value that encapsulate _both_ the "primary mode" and how the different access values will be translated into `yes|restricted|no`. So in retrospect it was confusing to conflate the "access category" with "primary mode". Hopefully most of this is documented in the comments.

Reply:
> Fundamentally the "access category" is a data abstraction that keeps the code simple and flexible.

I think it is not a good idea to invent such concepts as parameters of implementing styling logic, especially if their meaning changes when changing the design logic.

The primary concern when implementing this should be that other style developers can easily understand and modify the tag interpretation logic.  The secondary concern should be code performance since we are going to use it on the fly during rendering.  Code complexity itself is lower priority.

As i said - we actually want to make this distinction based on road class, because it depends on the road class and its line signature if we can only show a two class access classification or if we can show three classes.

Reply:
We can call it road_class rather than access type/category. That's all it is really. `highway=cycleway` and `highway=path, bicycle=designated` belong to the same "road class" because the access marking works in the same way.

Reply:
Then i'd suggest to simply re-cast `highway=path` into `highway=cycleway/bridleway/footway` in SQL and work with that as parameter.  You essentially already have the function for that (`carto_path_primary_mode()`).  We currently do this re-casting in MSS code - but since you need it for the access restriction processing you can do it in SQL and then simplify the MSS code as a bonus.

You'd need an additional query nesting level for that or move the `carto_highway_int_access()` call to the outer query.

Reply:
> Then i'd suggest to simply re-cast `highway=path` into `highway=cycleway/bridleway/footway` in SQL

Yes, I was thinking along these lines - an effective `int_highway`. I am keen to collapse the different road types in one place, and will have to check that this works across the wider framework.